File: lib-core/p.m.dataserver.yahoo/src/org/aiotrade/platform/modules/dataserver/yahoo/YahooQuoteServer.java
/*
 * Copyright (c) 2006-2007, AIOTrade Computing Co. and Contributors
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 *  o Redistributions of source code must retain the above copyright notice, 
 *    this list of conditions and the following disclaimer. 
 *    
 *  o Redistributions in binary form must reproduce the above copyright notice, 
 *    this list of conditions and the following disclaimer in the documentation 
 *    and/or other materials provided with the distribution. 
 *    
 *  o Neither the name of AIOTrade Computing Co. nor the names of 
 *    its contributors may be used to endorse or promote products derived 
 *    from this software without specific prior written permission. 
 *    
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; 
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.aiotrade.platform.modules.dataserver.yahoo;

import java.awt.Image;
import java.awt.image.BufferedImage;
import java.io.BufferedReader;
import java.io.File;
import java.io.IOException;
import java.io.InputStream;
import java.io.InputStreamReader;
import java.net.HttpURLConnection;
import java.net.URL;
import java.text.DateFormat;
import java.text.ParseException;
import java.text.SimpleDateFormat;
import java.util.Calendar;
import java.util.Date;
import java.util.Locale;
import java.util.TimeZone;
import java.util.zip.GZIPInputStream;
import javax.imageio.ImageIO;
import org.aiotrade.lib.math.timeseries.Frequency;
import org.aiotrade.platform.core.dataserver.QuoteContract;
import org.aiotrade.platform.core.dataserver.QuoteServer;
import org.aiotrade.platform.core.sec.Market;
import org.aiotrade.platform.core.sec.Quote;

/**
 * This class will load the quote datas from data source to its data storage: quotes.
 * @TODO it will be implemented as a Data Server ?
 *
 * @author Caoyuan Deng
 */
public class YahooQuoteServer extends QuoteServer {

    // * "http://table.finance.yahoo.com/table.csv"
    private static String BaseUrl = "http://aiotrade.com/";
    private static String UrlPath = "aiodata/yq";
    private DateFormat dateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.US);
    private DateFormat dateFormat_old = new SimpleDateFormat("dd-MMM-yy", Locale.US);
    private QuoteContract contract;
    private boolean gzipped;

    protected boolean connect() {
        return true;
    }

    /**
     * Template:
     * http://table.finance.yahoo.com/table.csv?s=^HSI&a=01&b=20&c=1990&d=07&e=18&f=2005&g=d&ignore=.csv
     */
    protected void request() throws Exception {
        sourceCal.clear();

        contract = getCurrentContract();

        Date begDate = new Date();
        Date endDate = new Date();
        if (getFromTime() <= ANCIENT_TIME /* @todo */) {
            begDate = contract.getBegDate();
            endDate = contract.getEndDate();
        } else {
            sourceCal.setTimeInMillis(getFromTime());
            begDate = sourceCal.getTime();
        }

        sourceCal.setTime(begDate);
        int a = sourceCal.get(Calendar.MONTH);
        int b = sourceCal.get(Calendar.DAY_OF_MONTH);
        int c = sourceCal.get(Calendar.YEAR);

        sourceCal.setTime(endDate);
        int d = sourceCal.get(Calendar.MONTH);
        int e = sourceCal.get(Calendar.DAY_OF_MONTH);
        int f = sourceCal.get(Calendar.YEAR);

        StringBuilder urlStr = new StringBuilder(50);
        urlStr.append(BaseUrl).append(UrlPath);
        urlStr.append("?s=").append(contract.getSymbol());

        /** a, d is month, which from 0 to 11 */
        urlStr.append(
                "&a=" + a + "&b=" + b + "&c=" + c +
                "&d=" + d + "&e=" + e + "&f=" + f);

        urlStr.append("&g=d&ignore=.csv");

        URL url = new URL(urlStr.toString());

        System.out.println(url);

        if (url != null) {
            HttpURLConnection conn = (HttpURLConnection) url.openConnection();
            conn.setRequestProperty("Accept-Encoding", "gzip");
            conn.setAllowUserInteraction(true);
            conn.setRequestMethod("GET");
            conn.setInstanceFollowRedirects(true);
            conn.connect();

            String encoding = conn.getContentEncoding();
            if (encoding != null && encoding.indexOf("gzip") != -1) {
                gzipped = true;
            } else {
                gzipped = false;
            }
            setInputStream(conn.getInputStream());
        }
    }

    /**
     * @return readed time
     */
    protected long read() throws Exception {
        InputStream is = getInputStream();
        if (is == null) {
            return 0;
        }

        BufferedReader reader;
        if (gzipped) {
            reader = new BufferedReader(new InputStreamReader(new GZIPInputStream(is)));
        } else {
            reader = new BufferedReader(new InputStreamReader(is));
        }

        /** skip first line */
        String s = reader.readLine();

        /** token index */
        int iDateTime = 0;
        int iOpen = 1;
        int iHigh = 2;
        int iLow = 3;
        int iClose = 4;
        int iVolume = 5;
        int iAdjClose = 6;

        long newestTime = -Long.MAX_VALUE;
        resetCount();
        sourceCal.clear();
        while ((s = reader.readLine()) != null) {
            String[] items;
            items = s.split(",");

            if (items.length < 6) {
                break;
            }

            /**
             * !NOTICE
             * must catch the date parse exception, other wise, it's dangerous
             * for build a calendarTimes in MasterSer
             */
            Date date = null;
            try {
                date = dateFormat.parse(items[iDateTime].trim());
            } catch (ParseException ex1) {
                try {
                    date = dateFormat_old.parse(items[iDateTime].trim());
                } catch (ParseException ex2) {
                    continue;
                }
            }
            sourceCal.clear();
            sourceCal.setTime(date);

            long time = sourceCal.getTimeInMillis();
            if (time < getFromTime()) {
                continue;
            }

            // quote time is rounded to 00:00, we should adjust it to open time
            String symbol = contract.getSymbol();
            time += getMarket(symbol).getOpenTimeOfDay();

            Quote quote = borrowQuote();

            quote.setTime(time);
            quote.setOpen(Float.parseFloat(items[iOpen].trim()));
            quote.setHigh(Float.parseFloat(items[iHigh].trim()));
            quote.setLow(Float.parseFloat(items[iLow].trim()));
            quote.setClose(Float.parseFloat(items[iClose].trim()));
            quote.setVolume(Float.parseFloat(items[iVolume].trim()) / 100f);
            quote.setAmount(-1);
            quote.setClose_adj(Float.parseFloat(items[iAdjClose].trim()));

            if (quote.getHigh() * quote.getLow() * quote.getClose() == 0) {
                returnQuote(quote);
                continue;
            }

            getStorage(contract).add(quote);

            newestTime = Math.max(newestTime, time);
            countOne();
        }

        return newestTime;
    }

    protected long loadFromSource(long afterThisTime) {
        setFromTime(afterThisTime + 1);

        long loadedTime = getLoadedTime();
        if (!connect()) {
            return loadedTime;
        }
        try {
            request();
            loadedTime = read();
        } catch (Exception ex) {
            System.out.println("Error in loading from source: " + ex.getMessage());
        }

        return loadedTime;
    }

    public String getDisplayName() {
        return "Yahoo! Finance Internet";
    }

    public String getDefaultDateFormatString() {
        return "yyyy-mm-dd";
    }

    public byte getSourceSerialNumber() {
        return (byte) 1;
    }

    @Override
    public Frequency[] getSupportedFreqs() {
        return new Frequency[]{Frequency.DAILY};
    }

    @Override
    public Image getIcon() {
        BufferedImage img;
        try {
            img = ImageIO.read(new File("org/aiotrade/platform/modules/dataserver/basic/netbeans/resources/favicon_yahoo.png"));
        } catch (IOException e) {
            img = null;
        }
        return img;
    }

    @Override
    public TimeZone getSourceTimeZone() {
        return TimeZone.getTimeZone("America/New_York");
    }

    public Market getMarket(String symbol) {
        return GetMarket(symbol);
    }

    public static Market GetMarket(String symbol) {
        String[] tokens = symbol.split("\\.");
        if (tokens.length >= 2) {
            String marketSym = tokens[tokens.length - 1];
            if (marketSym.equalsIgnoreCase("L")) {
                return Market.LDSE;
            } else if (marketSym.equalsIgnoreCase("SS")) {
                return Market.SHSE;
            } else if (marketSym.equalsIgnoreCase("SZ")) {
                return Market.SZSE;
            }
        }

        return Market.NYSE;
    }
}

File: modules/p.m.dataserver.yahoo/src/main/scala/org/aiotrade/platform/modules/dataserver/yahoo/YahooQuoteServer.scala
/*
 * Copyright (c) 2006-2007, AIOTrade Computing Co. and Contributors
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without 
 * modification, are permitted provided that the following conditions are met:
 * 
 *  o Redistributions of source code must retain the above copyright notice, 
 *    this list of conditions and the following disclaimer. 
 *    
 *  o Redistributions in binary form must reproduce the above copyright notice, 
 *    this list of conditions and the following disclaimer in the documentation 
 *    and/or other materials provided with the distribution. 
 *    
 *  o Neither the name of AIOTrade Computing Co. nor the names of 
 *    its contributors may be used to endorse or promote products derived 
 *    from this software without specific prior written permission. 
 *    
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" 
 * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, 
 * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR 
 * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR 
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, 
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, 
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; 
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, 
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR 
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, 
 * EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */
package org.aiotrade.platform.modules.dataserver.yahoo

import java.awt.Image
import java.awt.image.BufferedImage
import java.io.{BufferedReader, File, IOException, InputStream, InputStreamReader}
import java.net.{HttpURLConnection, URL}
import java.text.{DateFormat, ParseException, SimpleDateFormat}
import java.util.{Calendar, Date, Locale, TimeZone}
import java.util.zip.GZIPInputStream
import javax.imageio.ImageIO
import org.aiotrade.lib.math.timeseries.TFreq
import org.aiotrade.lib.securities.{Market, Quote}
import org.aiotrade.lib.securities.dataserver.{QuoteContract, QuoteServer}

/**
 * This class will load the quote datas from data source to its data storage: quotes.
 * @TODO it will be implemented as a Data Server ?
 *
 * @author Caoyuan Deng
 */
object YahooQuoteServer {
  // * "http://table.finance.yahoo.com/table.csv"
  protected val BaseUrl = "http://aiotrade.com/"
  protected val UrlPath = "aiodata/yq"
  protected val dateFormat: DateFormat = new SimpleDateFormat("yyyy-MM-dd", Locale.US)

  def marketOf(symbol: String): Market = {
    symbol.split("\\.") match {
      case Array(head, market) => market.toUpperCase match {
          case "L" =>  Market.LDSE
          case "SS" => Market.SHSE
          case "SZ" => Market.SZSE
          case _ => Market.NYSE
        }
      case _ => Market.NYSE
    }
  }
}

class YahooQuoteServer extends QuoteServer {
  import YahooQuoteServer._

  private var contract: QuoteContract = _
  private var gzipped = false

  protected def connect: Boolean = true

  /**
   * Template:
   * http://table.finance.yahoo.com/table.csv?s=^HSI&a=01&b=20&c=1990&d=07&e=18&f=2005&g=d&ignore=.csv
   */
  @throws(classOf[Exception])
  protected def request: Unit = {
    val cal = Calendar.getInstance

    contract = currentContract match {
      case Some(x: QuoteContract) => x
      case _ => return
    }

    val (begDate, endDate ) = if (fromTime <= ANCIENT_TIME /* @todo */) {
      (contract.beginDate, contract.endDate)
    } else {
      cal.setTimeInMillis(fromTime)
      (cal.getTime, new Date)
    }

    cal.setTime(begDate)
    val a = cal.get(Calendar.MONTH)
    val b = cal.get(Calendar.DAY_OF_MONTH)
    val c = cal.get(Calendar.YEAR)

    cal.setTime(endDate)
    val d = cal.get(Calendar.MONTH)
    val e = cal.get(Calendar.DAY_OF_MONTH)
    val f = cal.get(Calendar.YEAR)

    val urlStr = new StringBuilder(50)
    urlStr.append(BaseUrl).append(UrlPath)
    urlStr.append("?s=").append(contract.symbol)

    /** a, d is month, which from 0 to 11 */
    urlStr.append("&a=" + a + "&b=" + b + "&c=" + c +
                  "&d=" + d + "&e=" + e + "&f=" + f)

    urlStr.append("&g=d&ignore=.csv")

    val url = new URL(urlStr.toString)

    println(url)

    if (url != null) {
      val conn = url.openConnection.asInstanceOf[HttpURLConnection]
      conn.setRequestProperty("Accept-Encoding", "gzip")
      conn.setAllowUserInteraction(true)
      conn.setRequestMethod("GET")
      conn.setInstanceFollowRedirects(true)
      conn.connect

      val encoding = conn.getContentEncoding
      gzipped = if (encoding != null && encoding.indexOf("gzip") != -1) {
        true
      } else false
            
      inputStream = conn.getInputStream match {
        case null => None
        case is => Some(is)
      }
    }
  }

  /**
   * @return readed time
   */
  @throws(classOf[Exception])
  protected def read: Long =  {
    val is = inputStream match {
      case None => return 0
      case Some(x) => x
    }

    val reader = if (gzipped) {
      new BufferedReader(new InputStreamReader(new GZIPInputStream(is)))
    } else {
      new BufferedReader(new InputStreamReader(is))
    }

    /** skip first line */
    val s = reader.readLine

    resetCount
    val storage = storageOf(contract)
    val symbol = contract.symbol
    val market = marketOf(contract.symbol)
    val timeZone = market.timeZone
    // * for daily quote, yahoo returns market's local date, so use market time zone
    val cal = Calendar.getInstance(timeZone)
    val dateFormat = dateFormatOf(timeZone)
    def loop(newestTime: Long): Long = reader.readLine match {
      case null => newestTime // break now
      case line => line.split(",") match {
          case Array(dateTimeX, openX, highX, lowX, closeX, volumeX, adjCloseX, _*) =>
            /**
             * !NOTICE
             * must catch the date parse exception, other wise, it's dangerous
             * for build a calendarTimes in MasterSer
             */
            try {
              val date = dateFormat.parse(dateTimeX.trim)
              cal.clear
              cal.setTime(date)
            } catch {
              case _: ParseException => loop(newestTime)
            }
                    
            var time = cal.getTimeInMillis
            if (time < fromTime) {
              loop(newestTime)
            }

            // quote time is rounded to 00:00, we should adjust it to open time
            time += market.openTimeOfDay

            val quote = borrowQuote

            quote.time   = time
            quote.open   = openX.trim.toFloat
            quote.high   = highX.trim.toFloat
            quote.low    = lowX.trim.toFloat
            quote.close  = closeX.trim.toFloat
            quote.volume = volumeX.trim.toFloat / 100f
            quote.amount = -1
            quote.close_adj = adjCloseX.trim.toFloat

            val newestTime1 = if (quote.high * quote.low * quote.close == 0) {
              returnQuote(quote)
              newestTime
            } else {
              storage += quote
              countOne
              Math.max(newestTime, time)
            }
                        
            loop(newestTime1)
          case _ => loop(newestTime)
        }
    }

    loop(-Long.MaxValue)
  }

  protected def loadFromSource(afterThisTime: Long): Long = {
    fromTime = afterThisTime + 1

    var loadedTime1 = loadedTime
    if (!connect) {
      return loadedTime1
    }
        
    try {
      request
      loadedTime1 = read
    } catch {
      case ex: Exception => ex.printStackTrace
    }

    loadedTime1
  }

  override def displayName: String = "Yahoo! Finance Internet"

  def defaultDateFormatPattern: String = "yyyy-MM-dd"

  def sourceSerialNumber: Byte = 1

  override def supportedFreqs: Array[TFreq] = {
    Array(TFreq.DAILY)
  }

  override def icon: Option[Image] = {
    val img = try {
      ImageIO.read(new File("org/aiotrade/platform/modules/dataserver/yahoo/resources/favicon_yahoo.png"))
    } catch {case _ => null}

    img match {
      case null => None
      case _ => Some(img)
    }
  }

  override def sourceTimeZone: TimeZone = TimeZone.getTimeZone("America/New_York")

  def marketOf(symbol: String): Market  = {
    YahooQuoteServer.marketOf(symbol)
  }

}



